### PR TITLE
Ignore Null values in DecoderResult and EncoderResult

### PR DESCRIPTION
**Pull request by ashhSAG93**

### CODE REVIEW

File: microservice/custom-decoders/src/main/java/com/cumulocity/microservice/customdecoders/api/model/DecoderResult.java [hunk 1]
```diff
@@ -25,6 +26,7 @@
 
 @NoArgsConstructor
 @Setter
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (bbyreddy) [inline microservice/custom-decoders/src/main/java/com/cumulocity/microservice/customdecoders/api/model/DecoderResult.java]: - [ ] Add static import for JsonInclude so that you don't need to refer to JsonInclude.Include.NON_NULL

File: microservice/custom-encoders/src/main/java/com/cumulocity/microservice/customencoders/api/model/EncoderResult.java [hunk 2]
```diff
@@ -7,12 +7,14 @@
...
 package com.cumulocity.microservice.customencoders.api.model;
 
+import com.fasterxml.jackson.annotation.JsonInclude;
 import lombok.*;
 
 import java.util.Map;
 
 @NoArgsConstructor
 @Data
+@JsonInclude(JsonInclude.Include.NON_NULL)
```
Comment (bbyreddy) [inline microservice/custom-encoders/src/main/java/com/cumulocity/microservice/customencoders/api/model/EncoderResult.java]: - [ ] Add static import for JsonInclude so that you don't need to refer to JsonInclude.Include.NON_NULL

File: microservice/custom-encoders/pom.xml [hunk 3]
```diff
@@ -3,10 +3,9 @@
          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <parent>
-        <artifactId>clients-java</artifactId>
+        <artifactId>microservice-dependencies</artifactId>
```
Comment (bbyreddy) [inline microservice/custom-encoders/pom.xml:6]: - [x] Why change this? Custom encoders don't depend on microservices.

File: microservice/custom-encoders/pom.xml [hunk 4]
```diff
@@ -18,15 +17,18 @@
     </properties>
 
     <dependencies>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <scope>provided</scope>
+        </dependency>
```
Comment (bbyreddy) [inline microservice/custom-encoders/pom.xml:24]: - [x] Why is this change required?

File: microservice/lpwan-custom-codec/src/test/java/com/cumulocity/microservice/lpwan/codec/encoder/model/LpwanEncoderResultTest.java [hunk 5]
```diff
@@ -27,6 +30,19 @@ public void fportShouldNotAppearTwiceOnSerialization() throws JsonProcessingExce
...
+    public void nullFieldsShouldBeIgnoredOnSerialization() throws JsonProcessingException {
+        LpwanEncoderResult lpwanEncoderResult = new LpwanEncoderResult();
+        lpwanEncoderResult.setMessage("\"EncoderInputData is missing mandatory fields: 'sourceDeviceId', 'sourceDeviceEui', 'manufacturer, model and/or supportedCommands', 'commandName'\"");
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        String outputJson = objectMapper.writeValueAsString(lpwanEncoderResult);
+        assertThat(outputJson, containsString("message"));
+        assertThat(outputJson, containsString("success"));
+        assertThat(outputJson, not(containsString("encodedCommand")));
+        assertThat(outputJson, not(containsString("properties")));
```
Comment (bbyreddy) [inline microservice/lpwan-custom-codec/src/test/java/com/cumulocity/microservice/lpwan/codec/encoder/model/LpwanEncoderResultTest.java]: - [ ] You should use Json compare. new JsonExpectationsHelper().assertJsonEqual() or something similar. String comparison doesn't give a good indication to the developer who tries to fix the test when it fails.

**Comment by c8y-kh4mell** [comment]
Just a comment for testing

**Review by bbyreddy — APPROVED** [review]
Well done, Ashwin. Thanks!

File: microservice/custom-decoders/src/test/java/com/cumulocity/microservice/customdecoders/api/model/DecoderResultTest.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+public class DecoderResultTest {
+    @Test
+    public void nullFieldsShouldBeIgnoredOnSerialization() throws Exception {
+        DecoderResult decoderResult = new DecoderResult();
+        decoderResult.setMessage("\"DecoderInputData is missing mandatory fields: 'sourceDeviceId', 'sourceDeviceEui', 'inputData', 'manufacturer and/or model'\"");
+        ObjectMapper objectMapper = new ObjectMapper();
+
+        String expectedJson = "{\"message\":\"\\\"DecoderInputData is missing mandatory fields: 'sourceDeviceId', 'sourceDeviceEui', 'inputData', 'manufacturer and/or model'\\\"\",\"success\":true}";
+        String outputJson = objectMapper.writeValueAsString(decoderResult);
+        assertEquals(objectMapper.readTree(expectedJson), objectMapper.readTree(outputJson));
```
Comment (ashhSAG93) [inline microservice/custom-decoders/src/test/java/com/cumulocity/microservice/customdecoders/api/model/DecoderResultTest.java:17]: When we use the JsonExpectationsHelper to compare the 2 JSONs, we get the NoClassDefError for org/json/JSONException. We didnt want to add the org.json dependency with the hardcoded version. If we add the java-client dependency that contains org.json, there are a lot of other dependencies that are bought in, that arent needed. So we have used an approach using Jackson ObjectMapper, that generates JsonNode for both the expected and actual JSON and does a deep compare of the 2 JSONs.

File: microservice/lpwan-custom-codec/pom.xml [hunk 7]
```diff
@@ -79,7 +77,6 @@
         <dependency>
             <groupId>org.mockito</groupId>
             <artifactId>mockito-inline</artifactId>
-            <version>2.13.0</version>
```
Comment (ashhSAG93) [inline microservice/lpwan-custom-codec/pom.xml:82]: Version compliant with parent artifact dependencies

File: microservice/custom-encoders/src/main/java/com/cumulocity/microservice/customencoders/api/model/EncoderResult.java [hunk 8]
```diff
@@ -7,12 +7,16 @@
 
 package com.cumulocity.microservice.customencoders.api.model;
 
+import com.fasterxml.jackson.annotation.JsonInclude;
 import lombok.*;
 
 import java.util.Map;
 
+import static com.fasterxml.jackson.annotation.JsonInclude.*;
```
Comment (ashhSAG93) [inline microservice/custom-encoders/src/main/java/com/cumulocity/microservice/customencoders/api/model/EncoderResult.java:15]: Static Import has been added

File: microservice/custom-decoders/src/test/java/com/cumulocity/microservice/customdecoders/api/model/DecoderResultTest.java (hunk 6 above)
Comment (bbyreddy) [reply microservice/custom-decoders/src/test/java/com/cumulocity/microservice/customdecoders/api/model/DecoderResultTest.java:17]: You can still include org.json dependency but only for test, that way it doesn't get included in the output jar/library.

File: microservice/custom-encoders/pom.xml (hunk 3 above)
Comment (ashhSAG93) [reply microservice/custom-encoders/pom.xml:6]: OK. So we can revert it back to how it was right?

Comment (ashhSAG93) [reply microservice/custom-encoders/pom.xml:6]: The reason we changed to microservice-dependencies is as follows:

microservice-dependencies is a pom that has only dependenciesManagement (deps to be used by child projects), whereas the clients-java has its own dependencies under the <dependencies> section, which get pulled when used as a parent in a child project. So if we examine the dependency tree using mvn dependency:tree, we see that more dependencies are pulled in when we use clients-java than when we use microservice-dependencies. When we try other approaches, we do not get the jackson dependency that is required for this change. Hence we bought it in line with the custom-decoders.

File: microservice/custom-encoders/pom.xml (hunk 4 above)
Comment (ashhSAG93) [reply microservice/custom-encoders/pom.xml:24]: This change is the effect of the microservice-dependency change, the reason for which is mentioned in the previous comment

**Review by miryalarahul — APPROVED** [review]
